sync/sync.go

```diff
@@ -311,30 +311,48 @@ Register:
 	return t.reloadAndVerify(s, replicaAddress, toClient)
 }
 
-func (t *Task) isRevisionCountAndChainSame(fromClient, toClient *replicaClient.ReplicaClient) (bool, error) {
+func (t *Task) isRevisionCountAndChainSame(fromClient, toClient *replicaClient.ReplicaClient) (bool, []string, error) {
 	rwReplica, err := fromClient.GetReplica()
 	if err != nil {
-		return false, err
+		return false, nil, err
 	}
 
 	curReplica, err := toClient.GetReplica()
 	if err != nil {
-		return false, err
+		return false, nil, err
 	}
 	logrus.Infof("RevisionCount of RW replica (%v): %v, Current replica (%v): %v",
 		fromClient.GetAddress(), rwReplica.RevisionCounter, toClient.GetAddress(),
 		curReplica.RevisionCounter)
 	logrus.Infof("RW replica chain: %v, cur replica chain: %v", rwReplica.Chain, curReplica.Chain)
+	// Snapshots created after the checkpoint are only being synced,
+	// since till the checkpoint all the replicas are assumed to be
+	// having same data. Checkpoints are created only when the controller
+	// is certain that till this snapshot all the replicas are in sync.
+	if curReplica.Checkpoint != "" {
+		for indx, snapshot := range curReplica.Chain {
+			if snapshot == curReplica.Checkpoint {
+				curReplica.Chain = curReplica.Chain[:indx]
+				break
+			}
+		}
+		for indx, snapshot := range rwReplica.Chain {
```

Comment:
a check is required that curReplica Checkpoint should exists in rwReplica Chain.. else throw error
Similar one required in VerifyRebuildReplica